Add possibility to manual db-cache invalidation

porunov opened the pull request:
This PR adds possibility to manually expire / invalidate db-cache entries.

Related to #3155

Signed-off-by: Oleksandr Porunov <alexandr.porunov@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

porunov commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/database/cache/KCVSCacheInvalidationService.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public void markVertexAsExpired(Long vertexId) {
+        StaticBuffer vertexIdKey = idManager.getKey(vertexId);
```
Good point. I actually forgot to check that myself. I will double check and post an update

porunov replied:
You are right. This doesn't work for `indexStore`. Probably we can invalidate `edgeStore` only and `indexStore` should be invalidated by `key` somehow else.

porunov replied:
I guess I will change the interface of `CacheInvalidationService.java` to be able to expire either `edgeStore` or `indexStore` explicitly. For `edgeStore` we will be able to provide `vertexId` but for `indexStore` we will probably need to specify correct keys. Also, I guess a method like: `forceClearFullCache` could be helpful as well.

porunov replied:
@li-boxuan I'm think about the next interface for `CacheInvalidationService` below. What do you think about it?
```
/**
 * Marks specific vertex as expired.
 * It will make sure that any retrieved properties and edges associated with this vertex will be invalidated in vertex cache.
 * @param vertexId vertex id to expire
 */
void markVertexAsExpiredInEdgeStore(Long vertexId);

/**
 * Marks specific key as expired.
 * It will make sure that any retrieved properties and edges associated with this key will be invalidated in vertex cache.
 * @param key key to expire
 */
void markKeyAsExpiredInEdgeStore(StaticBuffer key);

/**
 * Marks specific key as expired.
 * It will make sure that any retrieved data associated with this key will be invalidated in index cache.
 * @param key key to expire
 */
void markKeyAsExpiredInIndexStore(StaticBuffer key);

/**
 * Instead of waiting for a probabilistic invalidation it triggers all cached queries scan and invalidation in vertex cache.
 * This will remove any cached expired data.
 */
void forceClearExpiredKeysInEdgeStoreCache();

/**
 * Instead of waiting for a probabilistic invalidation it triggers all cached queries scan and invalidation in index cache.
 * This will remove any cached expired data.
 */
void forceClearExpiredKeysInIndexStoreCache();

/**
 * Instead of waiting for a probabilistic invalidation it triggers all cached queries scan and invalidation for both edge store cache and index store cache.
 * This will remove any cached expired data.
 */
void forceClearExpiredCache();

/**
 * Marks a vertex as expired in edge store cache and triggers force clear of expired cache (i.e. {@link #forceClearExpiredKeysInEdgeStoreCache()})
 * @param vertexId vertex id to invalidate
 */
void forceInvalidateVertexInEdgeStoreCache(Long vertexId);

/**
 * Marks vertices as expired in edge store cache and triggers force clear of expired cache (i.e. {@link #forceClearExpiredKeysInEdgeStoreCache()})
 * @param vertexIds vertex ids to invalidate
 */
void forceInvalidateVerticesInEdgeStoreCache(Iterable<Long> vertexIds);

/**
 * Clears edge store cache fully
 */
void clearEdgeStoreCache();

/**
 * Clears index store cache fully
 */
void clearIndexStoreCache();

/**
 * Clears both edge store cache and index store cache fully.  
 * It is the same as calling {@link #clearEdgeStoreCache()} and {@link #clearIndexStoreCache()}
 */
void clearDBCache();
```

li-boxuan replied:
Before commenting on that, I would like to understand more about the context. I am not sure how `CacheInvalidationService` will be used. Do you expect this to be used directly by users, or by JanusGraph internals?

porunov replied:
I would like to have both possibilities. I.e. I want to let users to expire / clear cache manually when they need it and also would like to add some invalidation strategy which can invalidate necessary cache records globally on mutation.
In this PR I want to have only first solution (I.e. for manual invalidation) and in the follow up PRs I would like to have a global invalidation strategy added

li-boxuan replied:
A global invalidation strategy makes sense to me. I am hesitant about exposing this functionality to end users, mainly because it is non-trivial to evict the index cache. For example, when users do `graph.getDBCacheInvalidationService().forceInvalidateVertices(Arrays.asList(vertexId1, vertexId2)`, they would likely take it for granted that the index cache for `vertexId1` and `vertexId2` is also invalidated.

Taking a step back, I am wondering when this would be ever useful for end users to directly call these APIs. The internal usage (global invalidation) for sure is valuable but not sure if those APIs shall be exposed to end users.

porunov replied:
> Taking a step back, I am wondering when this would be ever useful for end users to directly call these APIs. The internal usage (global invalidation) for sure is valuable but not sure if those APIs shall be exposed to end users.

I agree that this invalidation service might not be obvious for users and maybe we need to hide it somewhere but I would still like to give advance users to access manual cache invalidation even so it's risky when they don't know how to work with cache invalidation. The reason for that is if users have their own log implementation with their own custom logic then can trigger cache invalidation on different nodes depending on their use case.
For example, if users already have custom log mechanism which logs only specific mutated relations (and not all relations), they can effectively invalidate them on different nodes. 
In our use case we don't want to invalidate all vertices when their properties are changed but we want to invalidate only those vertices which are served for clients. Any other vertex change is irrelevant to us and we don't want to log it or invalidate it.
If you feel that this service shouldn't be exposed to users publicly, I'm OK hiding it under `StandardJanusGraph`, but I would like to have a possibility to retrieve that service somehow without using Reflection API.
I guess if users really need to access DBCacheInvalidationService they can cast their JanusGraph to StandardJanusGraph are access it. I.e.:
```
StandardJanusGraph standardJanusGraph = (StandardJanusGraph) janusGraph;
DBCacheInvalidationService invalidationService = standardJanusGraph.getDBCacheInvalidationService();
``` 
What do you think about it?

li-boxuan replied:
Got it, that sounds like a legitimate use case although very advanced. I am happy to have it exposed as public APIs, I was just hesitant about including it in the doc. If we write it in the doc, we need to be very careful and make sure people don't misuse it. Do you think JavaDoc itself would be sufficient?

porunov replied:
> Got it, that sounds like a legitimate use case although very advanced. I am happy to have it exposed as public APIs, I was just hesitant about including it in the doc. If we write it in the doc, we need to be very careful and make sure people don't misuse it. Do you think JavaDoc itself would be sufficient?

I think you are right. I will exclude it from the doc and I will try to improve JavaDoc with some examples maybe, so that users won't be confused about db-cache invalidation.

porunov commented:
@li-boxuan changed the interface, excluded that previously added section from JanusGraph doc, covered new methods with integration tests (`edgeStore` invalidation by keys, `indexStore` invalidation by keys, full invalidation of `edgeStore` and `indexStore`), added JavaDoc to `CacheInvalidationService`.
The only thing I didn't fully check is the first example I added into the JavaDoc of `CacheInvalidationService#markKeyAsExpiredInIndexStore(StaticBuffer)` but I will double check that example tomorrow.
Other than that, the PR is ready for review

porunov commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/database/cache/CacheInvalidationService.java:
```diff
@@ -0,0 +1,192 @@
+     * from processing transaction logs via `ChangeState` then it can be done using an example like below:
+     *
+     * <pre>
+     * IndexSerializer indexSerializer = (StandardJanusGraph) graph.getIndexSerializer();
+     * CacheInvalidationService invalidationService = graph.getDBCacheInvalidationService();
+     * changeState.getRelations(Change.ANY).forEach(janusGraphRelation -&gt; {
+     *     if(janusGraphRelation.isProperty() &amp;&amp; janusGraphRelation instanceof InternalRelation){
+     *         JanusGraphVertexProperty property = (JanusGraphVertexProperty) janusGraphRelation;
+     *         if(property.element() instanceof InternalVertex){
+     *             Collection&lt;IndexUpdate&gt; indexUpdates = indexSerializer.getIndexUpdates((InternalVertex) property.element(),
+     *                 Collections.singleton((InternalRelation) property));
+     *
+     *             for(IndexUpdate indexUpdate : indexUpdates){
+     *                 StaticBuffer keyToInvalidate = (StaticBuffer) indexUpdate.getKey();
+     *                 invalidationService.markKeyAsExpiredInIndexStore(keyToInvalidate);
+     *             }
+     *
+     *             invalidationService.forceClearExpiredKeysInIndexStoreCache();
+     *             invalidationService.forceInvalidateVertexInEdgeStoreCache(property.element().longId());
+     *         }
+     *     }
+     * });
```
This example only partially correct because we also need to:
- Check that key is actually a `StaticBuffer` (for composite index) and not a `String` (for mixed index).
- Check for edge index updates
- Check added / removed vertices

If I add all these checks to this example than it becomes too complex to read in JavaDoc. Instead of this I propose removing this example and when #3155 is finished to simply add references to the appropriate methods for the developers who are interested to read it.

porunov replied:
Updated JavaDoc

li-boxuan commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/cache/ExpirationKCVSCache.java:
```diff
@@ -151,6 +158,31 @@ public void invalidate(StaticBuffer key, List<CachableStaticBuffer> entries) {
+        clearExpiredCache(false);
+    }
+
+    private synchronized void clearExpiredCache(boolean withNewPenaltyCountdown) {
```
Nvm I saw your reply here: https://github.com/JanusGraph/janusgraph/pull/3198#issuecomment-1234012307 I guess we would create a follow-up issue for this?

porunov replied:
Yeah, I think we will need to open follow-up PRs for direct invalidation without having expire functionality at all, but I didn't check the complexity of such operations yet

porunov replied:
@li-boxuan I created a follow-up issue regarding cache here: #3218

porunov commented:
@li-boxuan would you like to re-review this PR or is it OK to proceed with the lazy consensus?

li-boxuan commented:
@porunov Sorry I don't have much bandwidth recently. I've reviewed it once before anyways, so I am OK to proceed with the lazy consensus.

porunov commented:
> @porunov Sorry I don't have much bandwidth recently. I've reviewed it once before anyways, so I am OK to proceed with the lazy consensus.

Okay, I think it's safe to proceed with the lazy consensus because this PR adds new functionality and doesn't change logic of existing functionality.
